File: router/hybrid.py
"""Hybrid dispatcher (implementation.md §2 Core Routing Policy, steps 4-6).

Full end-to-end path: route with Jev (System 1), then
  - system1_auto  -> return the routing decision; the host harness executes
                     (System-1 execution is the harness's normal flow)
  - system2       -> escalate to the GLM consumer (router/system2.py) with
                     the retrieved memory context + routing reasons

Logs end-to-end latency and total cost per dispatch (§7.5 metrics): the
System-2 call is linked to its routing decision by decision_id.

Usage:
  python3 router/hybrid.py "user request text"
"""
from __future__ import annotations

import json
import sys
import time
from pathlib import Path

ROUTER_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(ROUTER_DIR))
sys.path.insert(0, str(ROUTER_DIR.parent / "evals" / "harness"))

from router import route  # noqa: E402
import memory  # noqa: E402
from system2 import System2Error, execute  # noqa: E402
from jev_client import JevError  # noqa: E402


def dispatch(request_text: str, project_root: str = ".", use_memory: bool = True) -> dict:
    """Route, then execute per the decision. Returns end-to-end metrics."""
    t0 = time.monotonic()

    # Step 1-4: System-1 routing (decision_id links all downstream logs)
    routing = route(request_text, project_root=project_root, use_memory=use_memory)
    decision_id = routing["decision_id"]

    out = {
        "request": request_text,
        "decision_id": decision_id,
        "routing": {k: routing[k] for k in
                    ("decision", "needs_review", "intent", "intent_confidence",
                     "safe_noul", "complexity_score", "reasons")},
        "system2": None,
        "end_to_end_ms": None,
        "cost_usd": None,
    }

    # Step 5: System-2 escalation path
    if routing["decision"] == "system2":
        ctx = memory.retrieve_all(request_text, project_root) if use_memory else None
        try:
            s2 = execute(
                request_text,
                context=memory.format_context(ctx) if ctx else "",
                reasons=routing["reasons"],
                decision_id=decision_id,
            )
            out["system2"] = {
                "status": "ok",
                "model": s2["model"],
                "text": s2["text"],
                "latency_ms": s2["latency_ms"],
                "usage": s2["usage"],
            }
        except System2Error as e:
            out["system2"] = {"status": "error", "error": str(e)}

    routing_cost = routing.get("usage", {}).get("cost", 0)
    s2_cost = (out["system2"] or {}).get("usage", {}).get("cost", 0) if out["system2"] else 0
    out["cost_usd"] = round(routing_cost + s2_cost, 8)
    out["end_to_end_ms"] = round((time.monotonic() - t0) * 1000, 1)
    return out


if __name__ == "__main__":
    def _degrade(reason_kind: str, reason: str) -> None:
        """Defined unavailable status — JSON, never a traceback (adapter contract)."""
        try:
            print(json.dumps({"status": "unavailable", "reason_kind": reason_kind,
                              "reason": reason[:300],
                              "routing_decision": "unavailable"}, indent=2))
        except BrokenPipeError:
            pass

    req = " ".join(sys.argv[1:]) or "Refactor the router package into a cleaner module layout."
    try:
        result = dispatch(req)
        summary = {k: result[k] for k in ("decision_id", "end_to_end_ms", "cost_usd")}
        summary["routing_decision"] = result["routing"]["decision"]
        if result["system2"] and result["system2"]["status"] == "ok":
            summary["system2_model"] = result["system2"]["model"]
            summary["system2_latency_ms"] = result["system2"]["latency_ms"]
            summary["response_preview"] = result["system2"]["text"][:200]
        print(json.dumps(summary, indent=2))
    except JevError as e:
        _degrade("missing_api_key" if "API_KEY" in str(e).upper() else "provider_error",
                 str(e))
    except Exception as e:  # noqa: BLE001 — internal bugs degrade too, never traceback
        _degrade("internal_error", f"{type(e).__name__}: {e}")

File: router/hybrid.ts
/**
 * Hybrid dispatcher (implementation.md §2 Core Routing Policy, steps 4-6).
 *
 * TypeScript port of the original hybrid.py. Full end-to-end path: route with Jev
 * (System 1), then
 *   - system1_auto  -> return the routing decision; the host harness executes
 *                      (System-1 execution is the harness's normal flow)
 *   - system2       -> escalate to the GLM consumer (router/system2.ts) with
 *                      the retrieved memory context + routing reasons
 *
 * Logs end-to-end latency and total cost per dispatch (§7.5 metrics): the
 * System-2 call is linked to its routing decision by decision_id.
 *
 * Usage: npx tsx router/hybrid.ts "user request text"
 */
import { performance } from "node:perf_hooks";
import { route } from "./router.ts";
import { retrieveAll, formatContext } from "./memory.ts";
import { System2Error, execute } from "./system2.ts";
import { JevError, usageCost, pythonRound } from "../evals/harness/jev_client.ts";
import { pyDumps, pyDumpsIndent, tagPythonFloats, PyFloat, type Json, type JsonTable } from "../evals/harness/jev_policy.ts";

interface DispatchResult extends JsonTable {
  request: string;
  decision_id: string;
  routing: JsonTable;
  system2: JsonTable | null;
  end_to_end_ms: Json;
  cost_usd: Json;
}

/** Route, then execute per the decision. Returns end-to-end metrics. */
export async function dispatch(
  requestText: string,
  projectRoot = ".",
  useMemory = true,
): Promise<DispatchResult> {
  const t0 = performance.now();

  // Step 1-4: System-1 routing (decision_id links all downstream logs)
  const routing = await route(requestText, { projectRoot, useMemory });
  const decisionId = routing["decision_id"];

  const out: DispatchResult = {
    request: requestText,
    decision_id: decisionId,
    routing: {
      decision: routing["decision"] as Json,
      needs_review: routing["needs_review"] as Json,
      intent: routing["intent"] as Json,
      intent_confidence: new PyFloat(routing["intent_confidence"]),
      safe_noul: new PyFloat(routing["safe_noul"]),
      complexity_score: new PyFloat(routing["complexity_score"]),
      reasons: routing["reasons"] as Json,
    },
    system2: null,
    end_to_end_ms: null,
    cost_usd: null,
  };

  // Step 5: System-2 escalation path
  if (routing["decision"] === "system2") {
    const ctx = useMemory ? await retrieveAll(requestText, projectRoot) : null;
    try {
      const s2 = await execute(requestText, {
        context: ctx ? formatContext(ctx) : "",
        reasons: routing["reasons"] as string[],
        decisionId,
      });
      out["system2"] = {
        status: "ok",
        model: s2["model"] as Json,
        text: s2["text"] as string,
        latency_ms: s2["latency_ms"] as Json,
        usage: s2["usage"] as Json,
      };
    } catch (error) {
      if (error instanceof System2Error) {
        out["system2"] = { status: "error", error: error.message };
      } else {
        throw error;
      }
    }
  }

  const routingCost = usageCost(routing["usage"]);
  const s2Table = out["system2"] as JsonTable | null;
  const s2Usage = s2Table !== null ? s2Table["usage"] : null;
  const s2Cost = usageCost(s2Usage);
  const known = [routingCost, s2Cost].filter((c): c is number => c !== null);
  out["cost_usd"] = known.length > 0 ? new PyFloat(pythonRound(known.reduce((x, y) => x + y, 0), 8)) : null;
  out["end_to_end_ms"] = new PyFloat(pythonRound(performance.now() - t0, 1));
  return out;
}

function degrade(reasonKind: string, reason: string): void {
  /** Defined unavailable status — JSON, never a traceback (adapter contract). */
  try {
    process.stdout.write(
      pyDumpsIndent(
        tagPythonFloats({
          status: "unavailable",
          reason_kind: reasonKind,
          reason: reason.slice(0, 300),
          routing_decision: "unavailable",
        }),
      ) + "\n",
    );
  } catch {
    // BrokenPipeError equivalent: ignore
  }
}

async function main(): Promise<number> {
  const req = process.argv.slice(2).join(" ") || "Refactor the router package into a cleaner module layout.";
  try {
    const result = await dispatch(req);
    const summary: JsonTable = {
      decision_id: result["decision_id"] as Json,
      end_to_end_ms: result["end_to_end_ms"] as Json,
      cost_usd: result["cost_usd"] as Json,
    };
    summary["routing_decision"] = (result["routing"] as JsonTable)["decision"] as Json;
    const s2 = result["system2"] as JsonTable | null;
    if (s2 && s2["status"] === "ok") {
      summary["system2_model"] = s2["model"] as Json;
      summary["system2_latency_ms"] = s2["latency_ms"] as Json;
      summary["response_preview"] = (s2["text"] as string).slice(0, 200);
    }
    process.stdout.write(pyDumpsIndent(tagPythonFloats(summary)) + "\n");
    return 0;
  } catch (error) {
    if (error instanceof JevError) {
      degrade(error.message.toUpperCase().includes("API_KEY") ? "missing_api_key" : "provider_error", error.message);
    } else if (error instanceof System2Error) {
      degrade("provider_error", error.message);
    } else {
      degrade("internal_error", `${(error as Error).name}: ${(error as Error).message}`);
    }
    return 0;
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const code = await main();
  process.exitCode = code;
}
